Fili config provider interface

jacobtolar opened the pull request:
Adds the config provider interface (this is the second child of #104) to bard core. There are a couple FIXMEs in here that should probably be addressed. What @michael-mclawhorn was talking about would be a more extensive change, I think.

Because of github silliness, this actually includes #107 so you'll probably just want to look at the most recent commit (https://github.com/yahoo/fili/commit/37d48b66a797d897f308140433eb66a98e904d80 at the moment). I can rebase when/if things are merged.

jacobtolar commented:
rebased from master after #107 was merged

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/MakerDictionary.java:
```diff
@@ -0,0 +1,295 @@
+     * @return an instantiated object, or null
+     */
+    public static <T> T instantiateObjectFromArgs(Class<? extends T> type, Object... args) {
+
```
This is cool code. But it doesn't seem tightly bound to the concept of making Makers.  Maybe relocate the reflections into a Utility class.

jacobtolar replied:
Cool/horrifying, there's a fine line :) But you're right, it could obviously construct any type of object.

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/MetricConfiguration.java:
```diff
@@ -0,0 +1,56 @@
+ * FIXME: The isExcluded() flag is currently unused.
+ */
+public interface MetricConfiguration {
+
```
Perhaps this class doesn't need to know about temporary dictionaries.  Perhaps load everything into a single dictionary and then filter out things that need filtering as a clean up stage.

jacobtolar replied:
Hmm. I'll play around with it; I don't like how it works currently, either.

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/ConfigProvider.java:
```diff
@@ -0,0 +1,60 @@
+     * @return a mapping of base metric names to their definitions
+     */
+    ConfigurationDictionary<MetricConfiguration> getBaseMetrics();
+
```
What is the use of distinguishing here between base and derived metrics?

jacobtolar replied:
I can't remember 😄 

I believe in the metric parsing code (not here) there's some assumptions that require them to be separate (same with the temporary metric dictionary somewhere else). I'll see if I can refactor it a bit so that this distinction isn't necessary.

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/ConfiguredTableLoader.java:
```diff
@@ -0,0 +1,111 @@
+            Set<ApiMetricName> apiMetricNames = new HashSet<>();
+            for (String metric : logicalTable.getMetrics()) {
+                apiMetricNames.add(new ApiMetricName() {
+                    @Override
```
It might make sense to extract this ApiMetricName factory code into a method so as to make it more easily tested.  It's also not strictly necessary to give every ApiMetricName it's own anonymous class.

jacobtolar replied:
Whoops, I thought I did that already - thanks

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/MakerDictionary.java:
```diff
@@ -0,0 +1,295 @@
+            MetricDictionary dict,
+            MakerDictionary mDict,
+            Map<String, MakerConfiguration> makers,
+            boolean skipDefault
```
Is there a reason you're not using your custom class 'ConfigurationDictionary<T>' here instead of crafting a raw map?

jacobtolar replied:
No, except that I think I created `ConfigurationDictionary` after I wrote this code. Could be changed.

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/MakerDictionary.java:
```diff
@@ -0,0 +1,295 @@
+     * 3. If no className given, the Maker name specified (longSum), turned into a Maker class name
+     *    (LongSumMaker), inside the MetricMaker package
+     *
+     * @param suppliedMakerName The maker name the user specified
```
This is a good start.  Another option for number 3 would be to draw on a list of configured maker dictionaries from config.  Not needed in the first draft.

michael-mclawhorn commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/PhysicalTableConfiguration.java:
```diff
@@ -0,0 +1,31 @@
+ * A PhysicalTableConfiguration can create its PhysicalTableDefinition, given a dimension configuration.
+ */
+public interface PhysicalTableConfiguration {
+
```
Do you have any implementing code for this class? 
What's the set of assumptions about how the metrics will be provided and how the dimension map will be used?

jacobtolar replied:
Yeah, just not in this PR.

The implementation I have for the YAML config provider: 
* Takes metric names, dimension names, and time grain in the constructor. 

* In buildPhysicalTable, you get a map of actual constructed dimension config objects. So the `dimensionMap` is used to resolve dimension names to DimensionConfig objects so a `PhysicalTableDefinition` can be built.

```
      @Override
      public PhysicalTableDefinition buildPhysicalTable(ConfigurationDictionary<DimensionConfig> dimensionMap) {
 
          Set<DimensionConfig> dimSet = dimensions
                  .stream()
                  .filter(dimensionMap::containsKey)
                  .map(dimensionMap::get)
                  .collect(Collectors.toSet());
 
          return new PhysicalTableDefinition(name, grain, dimSet);
      }
```

michael-mclawhorn replied:
It looks like this interface is a a PhysicalDefinitionFactory.  But you're calling it a PhysicalTableConfiguration.
As best as I can infer, this is going to be backed by a 'smart bean', where the underlying bean is initialized from a configuration model (e.g. YAML file, xml, database query, etc), and this INTERFACE discusses using that model AS a factory.

Maybe this is a conceptual confusion, or maybe this is a naming issue.

If you have an object modeling a thing that provides name, grain, dim names, metric names, maybe that bean should be an explicit thing. (e.g. a PhysicalTableConfiguration) and there should be another thing that implements what you've described above pulling the parts out of the model object and building a definition out of it.

michael-mclawhorn replied:
So I've thought about this some more.
I'm actually frustrated by some of the assumptions in the BaseTableLoader which this code has been forced to adapt to.

Are you imagining that in your config model that the list of field names will be strings in the tableDefinitionConfig, or that they'll be implicitly loaded in some other way?

jacobtolar replied:
Hmm, some of the other classes are like this, too (e.g. the MetricConfiguration interface has a `build` method that returns a logical metric).

For physical tables, it would be straightforward to clear this up by exposing dimensions and remove this method. The `PhysicalTableDefinition` would be created by the consumer of this class (e.g. `ConfiguredTableLoader` or something like that).

For metrics, it's a little trickier. I didn't want to force use of the metric parser so I wanted implementers of `MetricConfiguration` to be able to specify how a metric is built. I guess I could split that out into a MetricConfiguration and some sort of MetricFactory.

michael-mclawhorn commented on fili-core/src/test/groovy/com/yahoo/bard/webservice/data/config/provider/FuzzyQueryMatcher.groovy:
```diff
@@ -0,0 +1,135 @@
+ * Utility class to validate that TemplateDruidQueries have the same structure and values.
+ *
+ * It's ok if inner objects don't always have the same names.
+ */
```
What are the situations where names are non-deterministic?  Filtered aggregations?

jacobtolar replied:
It could be that there *shouldn't* be any, but the code I've written currently behaves that way (that code isn't in this PR). 

In the code I've written to parse/create logical metrics it creates temporary names for inner values (just increments a counter). Maybe there's a better way to do this, but it's simple/works.

So, something like `real_views = 2 * half_views + 7` would name the final logical metric `real_views` but would assign some temporary name to the inner nested metrics (e.g. `2*half_views`).

michael-mclawhorn commented on fili-core/src/test/groovy/com/yahoo/bard/webservice/data/config/provider/MakerDictionarySpec.groovy:
```diff
@@ -0,0 +1,40 @@
+
+        expect:
+        maker instanceof ThetaSketchMaker
+    }
```
You could go further and verify that the size of the sketch is correct as well.

jacobtolar replied:
Sure.

MetaAshley commented:
This is an automatically generated message.

awolf@yahoo-inc.com is no longer with Yahoo! Inc.

Your message will not be forwarded.

If you have a sales inquiry, please email yahoosales@yahoo-inc.com and someone will follow up with you shortly.

If you require assistance with a legal matter, please send a message to legal-notices@yahoo-inc.com

Thank you!

jacobtolar commented:
To address at least most of the discussion thus far (thanks @michael-mclawhorn):

* The global temporary metric dictionary is gone, replaced by just building makers on the fly
* The distinction between base/derived metrics is gone. Up to the user to supply them in the right order.
* Changed everything in the config provider interface to be lists of objects with a 'name' parameter.

jacobtolar commented on fili-core/src/main/java/com/yahoo/bard/webservice/data/config/provider/ConfigBinderFactory.java:
```diff
@@ -0,0 +1,93 @@
+            LOG.info("Loading ConfigBinderFactory for type: {}", className);
+            Class<? extends ConfigProvider> providerClass = (Class<? extends ConfigProvider>) Class.forName(className);
+            Method build = providerClass.getDeclaredMethod("build", SystemConfig.class);
+            provider = (ConfigProvider) build.invoke(null, SYSTEM_CONFIG);
```
I guess it's not strictly necessary to pass in the system config here. The config provider can get at it if it wants it.

jacobtolar commented:
@michael-mclawhorn @cdeszaq 

I moved this under `examples/fili-yaml-config-example` and rebased to the latest master.

This breaks the metric parser PR a bit because it depends on the "metric maker dictionary" stuff being in fili-core.
